### PR TITLE
refactor(python)!: Instantiate Polars datatypes

### LINKED ISSUE #6163: Polars datatypes should be instantiated
### Problem description

In Python, we have some advanced magic going on behind the scenes when it comes to handling Polars datatypes. This has some drawbacks that may lead to confusing behaviour for users:

```python
import polars as pl
from datetime import datetime

df = pl.DataFrame({'a': [1], 'b': [datetime(2023, 1, 1)]})

isinstance(df['a'].dtype, pl.Int64)  # False?
assert df['a'].dtype == pl.Int64  # Okay this works

isinstance(df['b'].dtype, pl.Datetime)  # True
assert df['b'].dtype == pl.Datetime  # This also works

map = {
    pl.Int64: 'x',
    pl.Datetime: 'y',
}

map[df['a'].dtype]  # Returns 'x'
map[df['b'].dtype]  # KeyError!
map[pl.Datetime]  # Returns 'y'
map[pl.Datetime(time_unit='us')]  # KeyError!

```

This behaviour stems from the fact that some dtypes are instantiated objects (mostly the ones that take arguments like `Datetime` and `List`), while others are uninstantiated (like `Int64`).

I think this is actually quite easy to fix: we should always work with instantiated DataType objects. If users specify a dtype input for e.g. a `Series`, we will instantiate it. `Series.dtype` will always return an instantiated object. `isinstance` will always work.

I only see benefits of this approach. But I'm sure the current setup is in place for a reason. Happy to get some feedback on this proposal!

### PR DESCRIPTION
**Pull request by stinodego**
Closes #6163 

The way we handle data types on the Python side is a big mess and leads to a lot of unexpected behaviour and bugs. This PR will lay the groundwork to start improving this.

The aim of this PR is to make the minimal required changes to make things work with 'instantiated' data types (see linked issue). I am **not** trying to solve all spaghetti code in one go. In fact, some things will be uglier now, but hopefully it's clear how this will lead to better code in the long run.

Some of the implications of these changes:
* Users can still provide schemas by specifying either instantiated (`Datetime()`) or uninstantiated (`Datetime`) dtypes. The latter is just a shorthand for the first. Behind the scenes, these are 'normalized' to all be instantiated.
* This means datatypes like `Datetime` and `List` no longer have class variables - only instance variables. Defaults for these are specified in the constructor (e.g. `List` defaults to `List(Float64)`). The concept of a generic `List` dtype without an inner dtype no longer really exists in the same way as before.
* Datatypes will always be exposed to the user in an instantiated way - e.g. `dtypes` and `schema` properties now return instantiated datatypes. _(This is a breaking change)_
* Checking for certain datatypes is simpler and can be done with `isinstance` checks.

Feedback is welcome! A lot of work is left, but as I said, I cannot solve all spaghetti in one go.

### CODE REVIEW

File: py-polars/polars/datatypes/classes.py [hunk 1]
```diff
@@ -4,47 +4,49 @@
 from inspect import isclass
 from typing import TYPE_CHECKING, Any, Iterator, Mapping, Sequence
 
-import polars.datatypes
+import polars.datatypes as dt
```
Comment (ghuls) [inline py-polars/polars/datatypes/classes.py:7]: Could this be imported with a different name? `dt` reminds me of datetime.

File: py-polars/polars/lazyframe/frame.py [hunk 2]
```diff
@@ -615,7 +618,7 @@ def schema(self) -> SchemaDict:
         {'foo': Int64, 'bar': Float64, 'ham': Utf8}
 
         """
-        return self._ldf.schema()
```
Comment (ghuls) [inline py-polars/polars/lazyframe/frame.py:618]: It would be better that this `_normalize_polars_dtype` would not be needed for dtypes and schema, but that the rust code for schema handles this directly:
https://github.com/stinodego/polars/blob/43c9def930a3c85511e02cbde02e525269f7e541/py-polars/src/lazy/dataframe.rs#L959-L969

Comment (stinodego) [reply py-polars/polars/lazyframe/frame.py:618]: I agree, but this is not trivial. The datatypes are an Enum in Rust, and the actual code for converting to Python classes is here:
https://github.com/pola-rs/polars/blob/61ce6f44d04570f654ad3b0586bfdf15cda045d9/py-polars/src/conversion.rs#L707-L807

Currently, there are some Python utils used in the Rust bindings to create the complex datatypes. I'd have to introduce a util for the other types. Which I can do, but that is also kind of ugly.

I think it might be preferable to define the datatype classes as PyO3 classes. But I'd have to take some time to look into that more closely. For now I think the current solution is fine.

File: py-polars/polars/datatypes/classes.py (hunk 1 above)
Comment (stinodego) [reply py-polars/polars/datatypes/classes.py:7]: Fair enough, we could do `pdt`.

File: py-polars/polars/lazyframe/frame.py (hunk 2 above)
Comment (ritchie46) [reply py-polars/polars/lazyframe/frame.py:618]: This is the code for converting a python object to a rust anyvalue. The code for converting to python dtypes is much simpler.

**Comment by ritchie46** [comment]
I really want the public API to have the illusion to work with `pl.Int32`.

Maybe we can instantiate objects in our public api: `Int32 = pl.datatypes.Int32()` and only expose those?

**Comment by stinodego** [comment]
> I really want the public API to have the illusion to work with pl.Int32.

With this PR, it still does, right? You can still input `pl.Int32` as before, it's handled just fine. The only change is that when calling `Series.dtype`, you get back `pl.Int32()`. That change is desirable, as it makes dtypes predictable and consistent to work with. If you disagree with that premise, we should probably close this PR.

> Maybe we can instantiate objects in our public api: `Int32 = pl.datatypes.Int32()` and only expose those? I think we should pre-instantiate all dtypes once and then schema can pick the pre-instantiated ones. (maybe you do this already?).

For simple dtypes that would work. But for complex types it doesn't make sense. If we define `Datetime = pl.datatypes.Datetime()`, you lose the ability to define `Datetime("ns")`.

**Comment by ritchie46** [comment]
>  The only change is that when calling Series.dtype, you get back pl.Int32(). 

Ok, that change doesn't seem too bad. I agree that it is more in line of expectations. :+1: 

> For simple dtypes that would work. But for complex types it doesn't make sense. If we define Datetime = pl.datatypes.Datetime(), you lose the ability to define Datetime("ns").

Right, that makes sense. As long as we still allow passing uninitialized types, I am all for this. :+1:

File: py-polars/polars/lazyframe/frame.py (hunk 2 above)
Comment (stinodego) [reply py-polars/polars/lazyframe/frame.py:618]: Hm, I may have been looking at the wrong thing then... let me look into this a bit more closely.

**Comment by alexander-beedie** [comment]
> For simple dtypes that would work. But for complex types it doesn't make sense. If we define `Datetime = pl.datatypes.Datetime()`, you lose the ability to define `Datetime("ns")`.

Actually there's a really easy solution for that 😅 
Just define `__call__` on the DataType.

**Comment by stinodego** [comment]
Sure, but the whole point of this change is to make things work more like a user would expect.

If I call `pl.Int32`, I expect to get a class, not an object.
If I call `pl.Int32()`, I expect to get an object, not a class.

We can do all kinds of magic, but let's just write dumb code that works?

**Comment by alexander-beedie** [comment]
Not sure how that's any stranger than having `pl.Datetime` and `pl.Datetime('us')` but disallowing `pl.Datetime()`? I'd argue that it's actually _more_ consistent because all dtypes appear callable instead of only some being special-cased  ;)

**Comment by stinodego** [comment]
The code in this PR allows for `pl.Datetime()` just fine - it will instantiate it with the defaults. Not sure we're understanding each other correctly here 😕

**Comment by stinodego** [comment]
As an aside - no need to rush this through for `0.17.0`. I think it offers a good basis to start improving some of our typing inconsistencies, but if we're not all on the same page we can hold off on that.

There are a lot of other approaches out there - the way we do it is pretty close to what pyspark does, but pyarrow and numpy have a completely different approach and also look pretty nice.

Maybe there needs to be a more comprehensive proposal before making any changes here.

**Comment by alexander-beedie** [comment]
Haha... maybe not; I probably concede things a bit by illustrating with `pl.Datetime`. Let me sanity check that we're on the same wavelength; with `__call__` _all_ of the following would work:

`pl.Int32`
`pl.Int32()`
`pl.Datetime` 
`pl.Datetime()` 
`pl.Datetime('us)` 

My thinking is that was thinking that you didn't want `pl.Int32()` to work? But if that's the case and `pl.Datetime()` does work then it seems inconsistent for the caller.

**Comment by stinodego** [comment]
There are various contexts in which we can use datatypes. There are two main cases I am interested in:

1. Specify a schema for a `DataFrame`. I would argue this is the most common case for users.
2. Do certain checks on your data, e.g. "is column X of type Y?". This is less common for users, but VERY common for us developers.

The first case remains unchanged, both in your proposal and in this PR. You specify types in exactly the same way.

However, the main issue with the current main branch is with case 2, as illustrated in the linked issue. If I am explicitly working with types and doing certain checks, they are wonky and behave unexpectedly.

This PR solves that completely (although not yet very elegantly).

Your proposal again gives me unexpected behaviour. As I said:

If I call pl.Int32, I expect to get a class, not an object. -> This is not the case for your proposal
If I call pl.Int32(), I expect to get an object, not a class. -> This is not the case for the current main branch, but is fixed with your proposal

In your proposal, you're just switching the problems around. This PR actually fixes this.

**Comment by alexander-beedie** [comment]
> e.g. "is column X of type Y?". This is less common for users, but VERY common for us developers.

Hmm. I think we're still missing each other... Running a quick regex search across the codebase shows that there are close to zero calls using `isinstance(x, SomeSpecificDType)` - the only one that I spotted is `isinstance(data_type, (DataType,DataTypeClass))` which would be unaffected by either approach - and use of `isinstance(x, SomeSpecificDType.base_type())` would solve pretty generically as that _always_ returns the uninitialised underlying DataType class, and is available from both the class and instance.

On the other hand there are quite a _lot_ of checks in the codebase using `x == SomeSpecificDType` or `x in (Tp1,Tp2,Tp3)` - which we already have this working with `Datetime` (as class or instance) via `__eq__`, and that would be trivial to have working with other instances (eg: a potential `Int32()`).

In other words... I really don't think there's a new problem being introduced this way, though it does indeed switch around one of the class/instance expectations. Anyway, not a huge deal, though this way does feel clean to me - possibly some bias from having already done it like that here 🤣 Still, either approach is a step forward 👍

**Comment by stinodego** [comment]
> there are close to zero calls using isinstance

That's because they don't work 😄 in this PR you can see that I replaced a lot of checks by `isinstance` calls.

The logic that `pl.Datetime("ms") == pl.Datetime` is actually confusing and shouldn't be true. An object is not equal to a class. But I left this in for now in order not to break too many things at once.

**Comment by stinodego** [comment]
I've given this some thought, and I am not fully confident in merging this right now. Test coverage for various datatype inputs are scarce, and I'm afraid of introducing unwanted behaviour. And I'll have limited time to address these in the coming weeks.

So I'll close this for now, and spend the coming weeks filing some issues for the datatype bugs I've found, improving test coverage, and improving the parsing/constructor logic. I'll also look into some other approaches for datatypes that I find promising like pyarrow's type system.

Then hopefully we can make the necessary breaking changes with the next major release.

**Comment by ritchie46** [comment]
Thanks @stinodego. I agree that it is good to tread with caution here. Setting up the test architecture for that sounds like a good approach.
